src/utils.ts

```diff
@@ -0,0 +1,214 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+import moment = require('moment');
+
+
+/*
+ * Copy the contents of one object to another, recursively.
+ *
+ * http://stackoverflow.com/questions/12317003/something-like-jquery-extend-but-standalone
+ */
+export
+function extend(target: any, source: any): any {
+  target = target || {};
+  for (var prop in source) {
+    if (typeof source[prop] === 'object') {
+      target[prop] = extend(target[prop], source[prop]);
+    } else {
+      target[prop] = source[prop];
+    }
+  }
+  return target;
+}
+
+
+/*
+ * Get a uuid as a string.
+ *
+ * http://www.ietf.org/rfc/rfc4122.txt
+ */
+export
+function uuid(): string {
+  var s: string[] = [];
+  var hexDigits = "0123456789ABCDEF";
+  for (var i = 0; i < 32; i++) {
+    s[i] = hexDigits.substr(Math.floor(Math.random() * 0x10), 1);
+  }
+  s[12] = "4";  // bits 12-15 of the time_hi_and_version field to 0010
+  s[16] = hexDigits.substr((Number(s[16]) & 0x3) | 0x8, 1);  // bits 6-7 of the clock_seq_hi_and_reserved to 01
+
+  var uuid = s.join("");
+  return uuid;
+}
+
+
+/**
+ * Join a sequence of url components with '/'.
+ */
+export
+function urlPathJoin(...paths: string[]): string {
```

Comment:
I'm not convinced that gains us anything. Instead of thinking "an array of path strings", I see "an array of Path objects" then go follow the link the "Path" interface which resolves to just a string.